### PR TITLE
Don't promote transitive dependencies in shaded hazelcast-all

### LINKED ISSUE #17873: hazelcast-all 4.1 declares compile-time dependencies
See: https://mvnrepository.com/artifact/com.hazelcast/hazelcast-all/4.1

image

That's quite serious, a lot of projects use the hazelcast-all artifact.

### PR DESCRIPTION
**Pull request by kwart**
Fixes #17873
Fixes hazelcast/hazelcast-enterprise#3857
EE side PR: hazelcast/hazelcast-enterprise#3858

This PR fixes the Hazelcast zero transitive dependency policy violations. The PR also adds regression tests to verify that no unexpected `compile` scoped dependencies are deployed.

If developers need to run the new dependency verification tests alone, they may do the following:
```bash
mvn clean install -DskipTests
mvn failsafe:integration-test -Dit.test=DependencyReducedPom* -DfailIfNoTests=false
```

### CODE REVIEW

**Comment by jerrinot** [comment]
is it possible to add a test? to make sure it wont happen again in the future. e.g. with a plugin version change.

**Comment by kwart** [comment]
> is it possible to add a test? to make sure it wont happen again in the future. e.g. with a plugin version change.

I'll think about a possible test approach.

File: hazelcast-all/src/test/java/com/hazelcast/it/DependencyReducedPomIT.java [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+
+import com.hazelcast.test.HazelcastParallelClassRunner;
+import com.hazelcast.test.annotation.QuickTest;
+
+@RunWith(HazelcastParallelClassRunner.class)
+@Category(QuickTest.class)
+public class DependencyReducedPomIT {
+
+    @Test
+    public void testZeroCompileScopedDeps() throws Exception {
```
Comment (lazerion) [inline hazelcast-all/src/test/java/com/hazelcast/it/DependencyReducedPomIT.java:45]: @kwart @jerrinot I think we need to add similar fail safe to release automation, I will work on it, please let me know if you see that redundant after this test case.

**Review by lazerion — APPROVED** [review]

File: hazelcast-all/src/test/java/com/hazelcast/it/DependencyReducedPomIT.java (hunk 1 above)
Comment (lazerion) [reply hazelcast-all/src/test/java/com/hazelcast/it/DependencyReducedPomIT.java:45]: @kwart I also see below when we execute mvn deploy

Comment (kwart) [reply hazelcast-all/src/test/java/com/hazelcast/it/DependencyReducedPomIT.java:45]: > I think we need to add similar fail safe to release automation, I will work on it, please let me know if you see that redundant after this test case.

We should ask @devozerov for an explanation. I think the 2 different shaded artifacts are intentional, but I didn't dig into it to get the reasoning.

**Comment by kwart** [comment]
I've added another commit which moves the `DocumentBuilderFactory` creation to the `XmlUtil` class.

**Comment by hz-highfive** [comment]
The job `Hazelcast-pr-builder` of your PR failed ([log](
https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/2971/build_log.txt), [artifacts](https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/2971/Hazelcast-pr-builder-2971.zip)). 
Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.
<details>
<summary>Click to expand the log file</summary>
<pre>
--------------------------
---------SUMMARY----------
--------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-checkstyle-plugin:3.1.1:checkstyle (default) on project hazelcast: An error has occurred in Checkstyle report generation. Failed during checkstyle execution: There is 1 error reported by Checkstyle 8.29 with /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/checkstyle/checkstyle.xml ruleset. -> [Help 1]
--------------------------
---------ERRORS-----------
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/hazelcast/src/test/java/com/hazelcast/it/DependencyReducedPomIT.java:1: Line does not match expected header line of '/*'. [Header]
--------------------------
</details>
</pre>

File: hazelcast-all/src/test/java/com/hazelcast/it/DependencyReducedPomIT.java (hunk 1 above)
Comment (devozerov) [reply hazelcast-all/src/test/java/com/hazelcast/it/DependencyReducedPomIT.java]: In the SQL module we have two executions of the `shade` module:
1. `fat-shaded-jar` - the default Jar with all dependencies merged into the `hazelcast-sql.jar`, this is what we publish 
1. `slim-original-jar` (previously `fat-jar`) produces a JAR that is convenient for the Jet team development process, it is not published (@gierlachg may provide additional details, my explanation could be inaccurate)

**Review by devozerov — APPROVED** [review]

File: hazelcast-all/src/test/java/com/hazelcast/it/DependencyReducedPomIT.java (hunk 1 above)
Comment (gierlachg) [reply hazelcast-all/src/test/java/com/hazelcast/it/DependencyReducedPomIT.java]: Hmm, why it is not published? We rely on it.

**Review by sancar — APPROVED** [review]
👍  for the changes in the client package.

**Comment by kwart** [comment]
Thank you for your reviews!

**Pull request merged**